File: model/avr25d/server/protocol.py
"""FrameMessage encode / decode — FROZEN after Day 1.

Wire layout (little-endian):
    [4 bytes]           header_len  — uint32, byte length of the JSON header
    [header_len bytes]  UTF-8 JSON  — the FrameMessage dict; typed-array fields
                                      carry the string "payload" as a sentinel
    [payload bytes...]  typed arrays in declaration order (cells, then refined):
                            cell_id   uint32[n]
                            ring      uint16[n]
                            bin       uint16[n]
                            z_ground  float32[n]
                            z_obstacle float32[n]
                            roughness  float32[n]
                            slope      float32[n]
                            class_id   uint8[n]
                            confidence uint8[n]
                            flags      uint8[n]
                        then refined (same layout, m entries):
                            parent_id  uint32[m]
                            quadrant   uint8[m]
                            z_ground   float32[m]
                            z_obstacle float32[m]
                            class_id   uint8[m]
                            flags      uint8[m]

Tracks, decision, and stats are fully JSON — they are small and the browser
reads them with zero-copy (JSON.parse is not on the hot path).

The browser reconstructs typed arrays from ArrayBuffer views over the payload
bytes that follow the header.  lib/protocol.ts mirrors this layout exactly.

Changing anything here after Day 1 requires the integration lead's sign-off
and a message to the whole team (IMPLEMENTATION_PLAN.md §5.1).
"""

from __future__ import annotations

import json
import struct
from dataclasses import asdict, dataclass, field
from typing import Any

import numpy as np

# ---------------------------------------------------------------------------
# dtype map: field name -> (numpy dtype, itemsize)
# ---------------------------------------------------------------------------

# Dense cells — in transmission order
_CELL_FIELDS: tuple[tuple[str, np.dtype], ...] = (
    ("cell_id",    np.dtype("<u4")),   # uint32
    ("ring",       np.dtype("<u2")),   # uint16
    ("bin",        np.dtype("<u2")),   # uint16
    ("z_ground",   np.dtype("<f4")),   # float32
    ("z_obstacle", np.dtype("<f4")),
    ("roughness",  np.dtype("<f4")),
    ("slope",      np.dtype("<f4")),
    ("class_id",   np.dtype("u1")),    # uint8
    ("confidence", np.dtype("u1")),
    ("flags",      np.dtype("u1")),
)

# Refined overlay — in transmission order
_REFINED_FIELDS: tuple[tuple[str, np.dtype], ...] = (
    ("parent_id",  np.dtype("<u4")),
    ("quadrant",   np.dtype("u1")),
    ("z_ground",   np.dtype("<f4")),
    ("z_obstacle", np.dtype("<f4")),
    ("class_id",   np.dtype("u1")),
    ("flags",      np.dtype("u1")),
)


# ---------------------------------------------------------------------------
# Thin containers — used by both Python sides (server and tests)
# ---------------------------------------------------------------------------

@dataclass
class CellArrays:
    """Dense occupied-cell data for one frame."""
    n: int
    cell_id:    np.ndarray   # uint32[n]
    ring:       np.ndarray   # uint16[n]
    bin:        np.ndarray   # uint16[n]
    z_ground:   np.ndarray   # float32[n]
    z_obstacle: np.ndarray   # float32[n]
    roughness:  np.ndarray   # float32[n]
    slope:      np.ndarray   # float32[n]
    class_id:   np.ndarray   # uint8[n]
    confidence: np.ndarray   # uint8[n]
    flags:      np.ndarray   # uint8[n]

    @staticmethod
    def empty() -> "CellArrays":
        return CellArrays(
            n=0,
            cell_id=np.zeros(0, "<u4"),
            ring=np.zeros(0, "<u2"),
            bin=np.zeros(0, "<u2"),
            z_ground=np.zeros(0, "<f4"),
            z_obstacle=np.zeros(0, "<f4"),
            roughness=np.zeros(0, "<f4"),
            slope=np.zeros(0, "<f4"),
            class_id=np.zeros(0, "u1"),
            confidence=np.zeros(0, "u1"),
            flags=np.zeros(0, "u1"),
        )


@dataclass
class RefinedArrays:
    """Sub-cell overlay (FR-17)."""
    n: int
    parent_id:  np.ndarray   # uint32[m]
    quadrant:   np.ndarray   # uint8[m]
    z_ground:   np.ndarray   # float32[m]
    z_obstacle: np.ndarray   # float32[m]
    class_id:   np.ndarray   # uint8[m]
    flags:      np.ndarray   # uint8[m]

    @staticmethod
    def empty() -> "RefinedArrays":
        return RefinedArrays(
            n=0,
            parent_id=np.zeros(0, "<u4"),
            quadrant=np.zeros(0, "u1"),
            z_ground=np.zeros(0, "<f4"),
            z_obstacle=np.zeros(0, "<f4"),
            class_id=np.zeros(0, "u1"),
            flags=np.zeros(0, "u1"),
        )


@dataclass
class Track:
    id: int
    x: float
    y: float
    vx: float
    vy: float
    class_id: int
    age: int
    speed: float
    predicted: list[list[float]] = field(default_factory=list)


@dataclass
class Decision:
    route: list[list[float]]
    alternative: list[list[float]]
    selected: str      # "primary" | "alternative"
    risk: str          # "LOW" | "MEDIUM" | "HIGH"
    eta_s: float
    reason: str


@dataclass
class FrameStats:
    fps: float
    t_perception_ms: float
    t_projection_ms: float
    t_analysis_ms: float
    t_refine_ms: float
    t_decision_ms: float
    t_serialise_ms: float
    t_total_ms: float
    n_points: int
    n_points_conserved: int   # FR-10: must equal n_points
    n_cells_occupied: int
    n_cells_total: int        # 705_771
    mem_bytes: int
    baseline_mem_bytes: int
    reduction: float


@dataclass
class FrameMessage:
    """One pipeline output frame, ready to encode."""
    frame_id:  int
    t_sec:     float
    mode:      str            # "live" | "cached" | "geometric"
    cells:     CellArrays
    refined:   RefinedArrays
    tracks:    list[Track]
    decision:  Decision
    stats:     FrameStats


# ---------------------------------------------------------------------------
# Encode
# ---------------------------------------------------------------------------

def encode(msg: FrameMessage) -> bytes:
    """Serialise a FrameMessage to the binary wire format.

    Returns: header_len (4 bytes, uint32 LE) | JSON header | payload bytes.
    """
    # Build the payload first so we can reference byte offsets in the header.
    payload_parts: list[bytes] = []

    def _append_arrays(arrays_data: list[tuple[str, np.ndarray]]) -> dict[str, str]:
        """Serialise arrays and return JSON sentinel dict."""
        sentinel: dict[str, str] = {}
        for name, arr in arrays_data:
            b = np.asarray(arr).tobytes()
            payload_parts.append(b)
            sentinel[name] = f"{arr.dtype}[{len(arr)}]"
        return sentinel

    cells_sentinel = _append_arrays([
        ("cell_id",    msg.cells.cell_id.astype("<u4")),
        ("ring",       msg.cells.ring.astype("<u2")),
        ("bin",        msg.cells.bin.astype("<u2")),
        ("z_ground",   msg.cells.z_ground.astype("<f4")),
        ("z_obstacle", msg.cells.z_obstacle.astype("<f4")),
        ("roughness",  msg.cells.roughness.astype("<f4")),
        ("slope",      msg.cells.slope.astype("<f4")),
        ("class_id",   msg.cells.class_id.astype("u1")),
        ("confidence", msg.cells.confidence.astype("u1")),
        ("flags",      msg.cells.flags.astype("u1")),
    ])
    refined_sentinel = _append_arrays([
        ("parent_id",  msg.refined.parent_id.astype("<u4")),
        ("quadrant",   msg.refined.quadrant.astype("u1")),
        ("z_ground",   msg.refined.z_ground.astype("<f4")),
        ("z_obstacle", msg.refined.z_obstacle.astype("<f4")),
        ("class_id",   msg.refined.class_id.astype("u1")),
        ("flags",      msg.refined.flags.astype("u1")),
    ])

    header_dict: dict[str, Any] = {
        "frame_id": msg.frame_id,
        "t_sec":    round(msg.t_sec, 4),
        "mode":     msg.mode,
        "cells": {"n": msg.cells.n, **cells_sentinel},
        "refined": {"n": msg.refined.n, **refined_sentinel},
        "tracks": [
            {
                "id":        t.id,
                "x":         round(t.x, 4),
                "y":         round(t.y, 4),
                "vx":        round(t.vx, 4),
                "vy":        round(t.vy, 4),
                "class_id":  t.class_id,
                "age":       t.age,
                "speed":     round(t.speed, 4),
                "predicted": [[round(p[0], 4), round(p[1], 4)] for p in t.predicted],
            }
            for t in msg.tracks
        ],
        "decision": {
            "route":       [[round(p[0], 4), round(p[1], 4)] for p in msg.decision.route],
            "alternative": [[round(p[0], 4), round(p[1], 4)] for p in msg.decision.alternative],
            "selected":    msg.decision.selected,
            "risk":        msg.decision.risk,
            "eta_s":       round(msg.decision.eta_s, 2),
            "reason":      msg.decision.reason,
        },
        "stats": {
            "fps":                  round(msg.stats.fps, 2),
            "t_perception_ms":      round(msg.stats.t_perception_ms, 2),
            "t_projection_ms":      round(msg.stats.t_projection_ms, 2),
            "t_analysis_ms":        round(msg.stats.t_analysis_ms, 2),
            "t_refine_ms":          round(msg.stats.t_refine_ms, 2),
            "t_decision_ms":        round(msg.stats.t_decision_ms, 2),
            "t_serialise_ms":       round(msg.stats.t_serialise_ms, 2),
            "t_total_ms":           round(msg.stats.t_total_ms, 2),
            "n_points":             msg.stats.n_points,
            "n_points_conserved":   msg.stats.n_points_conserved,
            "n_cells_occupied":     msg.stats.n_cells_occupied,
            "n_cells_total":        msg.stats.n_cells_total,
            "mem_bytes":            msg.stats.mem_bytes,
            "baseline_mem_bytes":   msg.stats.baseline_mem_bytes,
            "reduction":            round(msg.stats.reduction, 4),
        },
    }

    header_bytes = json.dumps(header_dict, separators=(",", ":")).encode("utf-8")
    header_len = struct.pack("<I", len(header_bytes))
    payload = b"".join(payload_parts)
    return header_len + header_bytes + payload


# ---------------------------------------------------------------------------
# Decode
# ---------------------------------------------------------------------------

def decode(data: bytes | bytearray | memoryview) -> dict[str, Any]:
    """Deserialise a FrameMessage from the wire format.

    Returns the header dict with typed-array sentinel strings intact
    and a ``_arrays`` sub-dict carrying the actual numpy arrays.

    The frontend (lib/protocol.ts) does the same reconstruction.
    """
    if len(data) < 4:
        raise ValueError(f"too short to contain a header length: {len(data)} bytes")

    header_len = struct.unpack_from("<I", data, 0)[0]
    if len(data) < 4 + header_len:
        raise ValueError(
            f"message truncated: need {4 + header_len} bytes, got {len(data)}"
        )

    header = json.loads(data[4:4 + header_len])
    payload = memoryview(data)[4 + header_len:]
    offset = 0

    def _read(dtype: np.dtype, n: int) -> np.ndarray:
        nonlocal offset
        nbytes = n * dtype.itemsize
        arr = np.frombuffer(payload[offset:offset + nbytes], dtype=dtype)
        offset += nbytes
        return arr

    n_cells = header["cells"]["n"]
    cells_arrays: dict[str, np.ndarray] = {}
    for name, dtype in _CELL_FIELDS:
        cells_arrays[name] = _read(dtype, n_cells)

    n_refined = header["refined"]["n"]
    refined_arrays: dict[str, np.ndarray] = {}
    for name, dtype in _REFINED_FIELDS:
        refined_arrays[name] = _read(dtype, n_refined)

    header["_arrays"] = {
        "cells":   cells_arrays,
        "refined": refined_arrays,
    }
    return header


# ---------------------------------------------------------------------------
# Round-trip helper (used by tests and fixtures)
# ---------------------------------------------------------------------------

def round_trip(msg: FrameMessage) -> dict[str, Any]:
    """encode then decode — used in tests to verify the wire contract."""
    return decode(encode(msg))

File: frontend/lib/protocol.ts
// protocol.ts — binary FrameMessage decode. Mirrors avr25d/server/protocol.py,
// which is FROZEN. If this file and that one disagree, this one is wrong.
//
// Wire layout (little-endian):
//   [4 bytes]           header_len — uint32
//   [header_len bytes]  UTF-8 JSON header; typed-array fields carry a sentinel
//                       string such as "uint32[41990]" in place of their data
//   [payload bytes...]  the typed arrays themselves, in declaration order:
//                         cells:   cell_id u32, ring u16, bin u16, z_ground f32,
//                                  z_obstacle f32, roughness f32, slope f32,
//                                  class_id u8, confidence u8, flags u8
//                         refined: parent_id u32, quadrant u8, z_ground f32,
//                                  z_obstacle f32, class_id u8, flags u8
//
// tracks, decision and stats live entirely in the JSON header — they are small
// and JSON.parse is not on the hot path.
//
// ── On "zero copies" ──────────────────────────────────────────────────────
// IMPLEMENTATION_PLAN.md §6.14 describes this decode as zero-copy. It cannot
// be, and the reason is arithmetic rather than opinion: a typed-array view
// needs its byte offset to be a multiple of its element size, the payload
// begins at 4 + header_len, and header_len is whatever the JSON happened to
// weigh. Measured against the live fixture server the payload starts at byte
// 1122 — 1122 % 4 == 2 — so a direct Uint32Array view over the received buffer
// throws RangeError on the very first frame, every frame.
//
// So the payload is copied once, contiguously, into its own ArrayBuffer. From
// offset 0 the cell arrays are self-aligning: cell_id at 0, ring at 4n, bin at
// 6n, the four float32 fields at 8n/12n/16n/20n — every offset a multiple of
// its element size for any n. Only `refined` can still land badly, because the
// cells block is 27n bytes and 27n % 4 == 0 only when n % 4 == 0; alignedView
// absorbs that case with a second, much smaller copy.
//
// One 1.1 MB memcpy per frame is far cheaper than the ~42,000 per-cell matrix
// writes the viewer already does downstream, and it is the only correct option
// without padding the frozen header. Raised with Anuj — see docs/progress/navya.md.

// ── Wire shapes ───────────────────────────────────────────────────────────
// These mirror the dataclasses in protocol.py. CellArrays is structurally
// identical to components/viewer/types.ts so the viewer consumes what this
// returns without a cast and without Shubham changing a line.

export interface CellArrays {
  n: number;
  cell_id: Uint32Array;
  ring: Uint16Array;
  bin: Uint16Array;
  z_ground: Float32Array;
  z_obstacle: Float32Array;
  roughness: Float32Array;
  slope: Float32Array;
  class_id: Uint8Array;
  confidence: Uint8Array;
  flags: Uint8Array;
}

/** Sub-cell overlay for far-field cells that were refined 2x2 (FR-17). */
export interface RefinedArrays {
  n: number;
  parent_id: Uint32Array;
  quadrant: Uint8Array;
  z_ground: Float32Array;
  z_obstacle: Float32Array;
  class_id: Uint8Array;
  flags: Uint8Array;
}

export interface Track {
  id: number;
  x: number;
  y: number;
  vx: number;
  vy: number;
  class_id: number;
  age: number;
  speed: number;
  /** Predicted future positions as [x, y] pairs. */
  predicted: number[][];
}

export interface Decision {
  route: number[][];
  alternative: number[][];
  selected: 'primary' | 'alternative';
  risk: 'LOW' | 'MEDIUM' | 'HIGH';
  eta_s: number;
  reason: string;
}

// A type alias, deliberately, NOT an interface. components/viewer/types.ts
// types stats as { [key: string]: number | string }, and TypeScript only gives
// implicit index signatures to type aliases — an interface would fail to
// satisfy it and the viewer would stop accepting decoded frames.
export type FrameStats = {
  fps: number;
  t_perception_ms: number;
  t_projection_ms: number;
  t_analysis_ms: number;
  t_refine_ms: number;
  t_decision_ms: number;
  t_serialise_ms: number;
  t_total_ms: number;
  n_points: number;
  /** FR-10: must equal n_points on every frame. */
  n_points_conserved: number;
  n_cells_occupied: number;
  n_cells_total: number;
  mem_bytes: number;
  baseline_mem_bytes: number;
  reduction: number;
};

export interface FrameMessage {
  frame_id: number;
  t_sec: number;
  /** FR-6 perception mode: "live" | "cached" | "geometric". */
  mode: string;
  cells: CellArrays;
  refined: RefinedArrays;
  tracks: Track[];
  decision: Decision;
  stats: FrameStats;
}

export class ProtocolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ProtocolError';
  }
}

// ── Field tables — order is the wire contract ─────────────────────────────

type Ctor =
  | Uint32ArrayConstructor
  | Uint16ArrayConstructor
  | Uint8ArrayConstructor
  | Float32ArrayConstructor;

/** numpy dtype name -> the view constructor, for validating the sentinels. */
const CELL_FIELDS: ReadonlyArray<readonly [keyof CellArrays, Ctor, string]> = [
  ['cell_id', Uint32Array, 'uint32'],
  ['ring', Uint16Array, 'uint16'],
  ['bin', Uint16Array, 'uint16'],
  ['z_ground', Float32Array, 'float32'],
  ['z_obstacle', Float32Array, 'float32'],
  ['roughness', Float32Array, 'float32'],
  ['slope', Float32Array, 'float32'],
  ['class_id', Uint8Array, 'uint8'],
  ['confidence', Uint8Array, 'uint8'],
  ['flags', Uint8Array, 'uint8'],
] as const;

const REFINED_FIELDS: ReadonlyArray<readonly [keyof RefinedArrays, Ctor, string]> = [
  ['parent_id', Uint32Array, 'uint32'],
  ['quadrant', Uint8Array, 'uint8'],
  ['z_ground', Float32Array, 'float32'],
  ['z_obstacle', Float32Array, 'float32'],
  ['class_id', Uint8Array, 'uint8'],
  ['flags', Uint8Array, 'uint8'],
] as const;

/** Bytes one cell occupies across all ten arrays — 27. */
const CELL_STRIDE = CELL_FIELDS.reduce((s, [, C]) => s + C.BYTES_PER_ELEMENT, 0);
/** Bytes one refined sub-cell occupies — 15 (4+1+4+4+1+1). */
const REFINED_STRIDE = REFINED_FIELDS.reduce((s, [, C]) => s + C.BYTES_PER_ELEMENT, 0);

// ── Reader ────────────────────────────────────────────────────────────────

/**
 * A view of `count` elements at `offset`, copying only if the offset is not a
 * multiple of the element size. After the payload slice this fires for the
 * refined block and nothing else.
 */
function alignedView<T extends Ctor>(
  Ctor: T,
  buffer: ArrayBuffer,
  offset: number,
  count: number
): InstanceType<T> {
  const bytes = count * Ctor.BYTES_PER_ELEMENT;
  if (offset + bytes > buffer.byteLength) {
    throw new ProtocolError(
      `payload truncated: need ${offset + bytes} bytes, have ${buffer.byteLength}`
    );
  }
  if (offset % Ctor.BYTES_PER_ELEMENT === 0) {
    return new Ctor(buffer, offset, count) as InstanceType<T>;
  }
  return new Ctor(buffer.slice(offset, offset + bytes)) as InstanceType<T>;
}

/**
 * Check a sentinel such as "uint32[41990]" against the field it stands for.
 * The protocol is frozen, so a mismatch means the backend moved and the right
 * response is a loud failure rather than a plausible-looking wrong picture.
 */
function checkSentinel(
  group: string,
  field: string,
  sentinel: unknown,
  dtype: string,
  n: number
): void {
  const expected = `${dtype}[${n}]`;
  if (sentinel !== expected) {
    throw new ProtocolError(
      `${group}.${field}: header says ${JSON.stringify(sentinel)}, expected "${expected}" ` +
        `— avr25d/server/protocol.py and lib/protocol.ts have diverged`
    );
  }
}

/**
 * Decode one binary FrameMessage.
 *
 * Returns `null` for a zero-length message: the server sends empty frames as
 * keepalives while the pipeline queue is drained (server/app.py), and treating
 * one as a frame throws on every startup.
 */
export function decodeFrame(data: ArrayBuffer): FrameMessage | null {
  if (data.byteLength === 0) return null;

  if (data.byteLength < 4) {
    throw new ProtocolError(
      `too short to contain a header length: ${data.byteLength} bytes`
    );
  }

  const headerLen = new DataView(data).getUint32(0, true); // little-endian
  const payloadStart = 4 + headerLen;
  if (data.byteLength < payloadStart) {
    throw new ProtocolError(
      `message truncated: need ${payloadStart} bytes of header, got ${data.byteLength}`
    );
  }

  const headerText = new TextDecoder('utf-8').decode(new Uint8Array(data, 4, headerLen));
  let header: Record<string, unknown>;
  try {
    header = JSON.parse(headerText);
  } catch (cause) {
    throw new ProtocolError(`header is not valid JSON: ${(cause as Error).message}`);
  }

  const cellsHeader = header.cells as Record<string, unknown>;
  const refinedHeader = header.refined as Record<string, unknown>;
  if (!cellsHeader || !refinedHeader) {
    throw new ProtocolError('header is missing the `cells` or `refined` section');
  }

  const n = cellsHeader.n as number;
  const m = refinedHeader.n as number;
  if (!Number.isInteger(n) || n < 0 || !Number.isInteger(m) || m < 0) {
    throw new ProtocolError(`bad array counts: cells.n=${n}, refined.n=${m}`);
  }

  const expectedBytes = n * CELL_STRIDE + m * REFINED_STRIDE;
  const actualBytes = data.byteLength - payloadStart;
  if (actualBytes < expectedBytes) {
    throw new ProtocolError(
      `payload truncated: ${n} cells + ${m} refined need ${expectedBytes} bytes, got ${actualBytes}`
    );
  }

  // The one copy. See the note at the top of this file.
  const payload = data.slice(payloadStart, payloadStart + expectedBytes);

  let offset = 0;
  const cells = { n } as CellArrays;
  for (const [field, Ctor, dtype] of CELL_FIELDS) {
    checkSentinel('cells', field, cellsHeader[field], dtype, n);
    // @ts-expect-error — field indexes a union of typed-array properties; the
    // FIELDS table is the thing that guarantees each name gets its own type.
    cells[field] = alignedView(Ctor, payload, offset, n);
    offset += n * Ctor.BYTES_PER_ELEMENT;
  }

  const refined = { n: m } as RefinedArrays;
  for (const [field, Ctor, dtype] of REFINED_FIELDS) {
    checkSentinel('refined', field, refinedHeader[field], dtype, m);
    // @ts-expect-error — as above.
    refined[field] = alignedView(Ctor, payload, offset, m);
    offset += m * Ctor.BYTES_PER_ELEMENT;
  }

  return {
    frame_id: header.frame_id as number,
    t_sec: header.t_sec as number,
    mode: header.mode as string,
    cells,
    refined,
    tracks: (header.tracks ?? []) as Track[],
    decision: header.decision as Decision,
    stats: header.stats as FrameStats,
  };
}

/** Bytes on the wire per occupied cell (27) — used by the HUD's throughput readout. */
export const BYTES_PER_CELL = CELL_STRIDE;
/** Bytes on the wire per refined sub-cell (15). */
export const BYTES_PER_REFINED_CELL = REFINED_STRIDE;
